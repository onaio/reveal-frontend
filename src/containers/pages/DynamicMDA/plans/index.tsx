import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import {
  HIDDEN_PLAN_STATUSES,
  SUPERSET_DYNAMIC_MDA_REPORTING_PLANS_SLICE,
} from '../../../../configs/env';
import { MDA_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_MDA_PLAN_URL } from '../../../../constants';
import { getPlanStatusToDisplay, getQueryParams, RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import GenericPLansReducer, {
  genericFetchPlans,
  makeGenericPlansArraySelector,
  reducerName as GenericPLansReducerName,
} from '../../../../store/ducks/generic/plans';
import { GenericPlan } from '../../../../store/ducks/generic/plans';
import { InterventionType } from '../../../../store/ducks/plans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the MDA plan definitions reducer */
reducerRegistry.register(GenericPLansReducerName, GenericPLansReducer);

/** a list of plan statuses to be displayed */
const allowedPlanStatusList = getPlanStatusToDisplay(HIDDEN_PLAN_STATUSES);

/** selector for Dynamic MDA plans */
const MDAPlansArraySelector = makeGenericPlansArraySelector();

/** Simple component that loads a list of MDA plans */
const MDAPLansList = (props: GenericPlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for MDAPLansList */
const defaultProps: GenericPlanListProps = {
  fetchPlans: genericFetchPlans,
  pageTitle: MDA_PLANS,
  pageUrl: REPORT_MDA_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_DYNAMIC_MDA_REPORTING_PLANS_SLICE,
};

MDAPLansList.defaultProps = defaultProps;

export { MDAPLansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: GenericPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const plans = MDAPlansArraySelector(state, {
    interventionTypes: [InterventionType.DynamicMDA],
    plan_title: searchedTitle,
    statusList: allowedPlanStatusList,
  });

  return {
    ...ownProps,
    plans,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: genericFetchPlans };

/** Connected MDA Plans component */
const ConnectedMDAPLansList = connect(mapStateToProps, mapDispatchToProps)(MDAPLansList);

export default ConnectedMDAPLansList;
