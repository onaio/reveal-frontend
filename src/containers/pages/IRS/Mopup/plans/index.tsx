import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { MOP_UP_REPORTING_TITLE } from '../../../../../configs/lang';
import { IRS_MOP_UP_REPORT_URL } from '../../../../../constants';
import { RouteParams } from '../../../../../helpers/utils';
import ConnectedIRSPlansList from '../../plans';

export const IRSMopUpReporting = (props: RouteComponentProps<RouteParams>) => {
  const plansProps = {
    ...props,
    pageTitle: MOP_UP_REPORTING_TITLE,
    pageUrl: IRS_MOP_UP_REPORT_URL,
  };
  return <ConnectedIRSPlansList {...plansProps} />;
};
