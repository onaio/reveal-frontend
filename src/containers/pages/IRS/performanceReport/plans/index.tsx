import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { IRS_PERFORMANCE_REPORTING_TITLE } from '../../../../../configs/lang';
import { PERFORMANCE_REPORT_IRS_PLAN_URL } from '../../../../../constants';
import { RouteParams } from '../../../../../helpers/utils';
import ConnectedIRSPlansList from '../../plans';

export const IRSPerfomenceReport = (props: RouteComponentProps<RouteParams>) => {
  const plansProps = {
    ...props,
    pageTitle: IRS_PERFORMANCE_REPORTING_TITLE,
    pageUrl: PERFORMANCE_REPORT_IRS_PLAN_URL,
  };
  return <ConnectedIRSPlansList {...plansProps} />;
};
