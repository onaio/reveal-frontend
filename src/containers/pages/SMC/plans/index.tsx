import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import { HIDDEN_PLAN_STATUSES, SUPERSET_SMC_REPORTING_PLANS_SLICE } from '../../../../configs/env';
import { SMC_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_SMC_PLAN_URL } from '../../../../constants';
import { getPlanStatusToDisplay, getQueryParams } from '../../../../helpers/utils';
import { RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import SMCPlansReducer, {
  fetchSMCPlans,
  makeSMCPlansArraySelector,
  reducerName as SMCPlansReducerName,
  SMCPLANType,
} from '../../../../store/ducks/generic/SMCPlans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the SMC plan definitions reducer */
reducerRegistry.register(SMCPlansReducerName, SMCPlansReducer);

/** a list of plan statuses to be displayed */
const allowedPlanStatusList = getPlanStatusToDisplay(HIDDEN_PLAN_STATUSES);
// store.dispatch(fetchSMCPlans(testlans as any))

/** Simple component that loads a preview list of SMC plans */
const SMCPlansList = (props: GenericPlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for SMCPlansList */
const defaultProps: GenericPlanListProps = {
  fetchPlans: fetchSMCPlans,
  pageTitle: SMC_PLANS,
  pageUrl: REPORT_SMC_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_SMC_REPORTING_PLANS_SLICE,
};

SMCPlansList.defaultProps = defaultProps;

export { SMCPlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: SMCPLANType[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const SMCPlansArray = makeSMCPlansArraySelector()(state, {
    plan_title: searchedTitle,
    statusList: allowedPlanStatusList,
  });

  return {
    ...ownProps,
    plans: SMCPlansArray,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: fetchSMCPlans };

/** Connected ActiveFI component */
const ConnectedSMCPlansList = connect(mapStateToProps, mapDispatchToProps)(SMCPlansList);

export default ConnectedSMCPlansList;
