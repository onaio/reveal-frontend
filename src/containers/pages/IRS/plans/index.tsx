import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import { HIDDEN_PLAN_STATUSES, SUPERSET_IRS_REPORTING_PLANS_SLICE } from '../../../../configs/env';
import { IRS_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_IRS_PLAN_URL } from '../../../../constants';
import { getPlanStatusToDisplay, getQueryParams, RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import GenericPLansReducer, {
  genericFetchPlans,
  GenericPlan,
  makeGenericPlansArraySelector,
  reducerName as GenericPLansReducerName,
} from '../../../../store/ducks/generic/plans';
import { InterventionType } from '../../../../store/ducks/plans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the IRS plan definitions reducer */
reducerRegistry.register(GenericPLansReducerName, GenericPLansReducer);

/** a list of plan statuses to be displayed */
const allowedPlanStatusList = getPlanStatusToDisplay(HIDDEN_PLAN_STATUSES);

/** selector for IRS plans */
const makeIRSPlansArraySelector = makeGenericPlansArraySelector();

/** Simple component that loads a preview list of IRS plans */
const IRSPlansList = (props: GenericPlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for IRSPlansList */
const defaultProps: GenericPlanListProps = {
  fetchPlans: genericFetchPlans,
  pageTitle: IRS_PLANS,
  pageUrl: REPORT_IRS_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_IRS_REPORTING_PLANS_SLICE,
};

IRSPlansList.defaultProps = defaultProps;

export { IRSPlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: GenericPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const plans = makeIRSPlansArraySelector(state, {
    interventionTypes: [InterventionType.IRS, InterventionType.DynamicIRS],
    plan_title: searchedTitle,
    statusList: allowedPlanStatusList,
  });

  return {
    ...ownProps,
    plans,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: genericFetchPlans };

/** Connected ActiveFI component */
const ConnectedIRSPlansList = connect(mapStateToProps, mapDispatchToProps)(IRSPlansList);

export default ConnectedIRSPlansList;
