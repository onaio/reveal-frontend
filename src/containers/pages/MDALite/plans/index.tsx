import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import {
  HIDDEN_PLAN_STATUSES,
  SUPERSET_MDA_LITE_REPORTING_PLANS_SLICE,
} from '../../../../configs/env';
import { MDA_LITE_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_MDA_LITE_PLAN_URL } from '../../../../constants';
import { getPlanStatusToDisplay, getQueryParams, RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import GenericPlansReducer, {
  genericFetchPlans,
  GenericPlan,
  makeGenericPlansArraySelector,
  reducerName as genericReducerName,
} from '../../../../store/ducks/generic/plans';
import { InterventionType } from '../../../../store/ducks/plans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the MDA plan definitions reducer */
reducerRegistry.register(genericReducerName, GenericPlansReducer);

/** a list of plan statuses to be displayed */
const allowedPlanStatusList = getPlanStatusToDisplay(HIDDEN_PLAN_STATUSES);

/** selector for MDA lite plans */
const makeMDALitePlansArraySelector = makeGenericPlansArraySelector();

/** Simple component that loads a preview list of MDA plans */
const MDALitePlansList = (props: GenericPlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for MDALitePlansList */
const defaultProps: GenericPlanListProps = {
  fetchPlans: genericFetchPlans,
  pageTitle: MDA_LITE_PLANS,
  pageUrl: REPORT_MDA_LITE_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_MDA_LITE_REPORTING_PLANS_SLICE,
};

MDALitePlansList.defaultProps = defaultProps;

export { MDALitePlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: GenericPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const MDALitePlansArray = makeMDALitePlansArraySelector(state, {
    interventionTypes: [InterventionType.MDALite],
    plan_title: searchedTitle,
    statusList: allowedPlanStatusList,
  });

  return {
    ...ownProps,
    plans: MDALitePlansArray,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: genericFetchPlans };

/** Connected ActiveFI component */
const ConnectedMDALitePlansList = connect(mapStateToProps, mapDispatchToProps)(MDALitePlansList);

export default ConnectedMDALitePlansList;
