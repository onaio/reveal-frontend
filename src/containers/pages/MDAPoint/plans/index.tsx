import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import {
  HIDDEN_PLAN_STATUSES,
  SUPERSET_MDA_POINT_REPORTING_PLANS_SLICE,
} from '../../../../configs/env';
import { MDA_POINT_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_MDA_POINT_PLAN_URL } from '../../../../constants';
import { getPlanStatusToDisplay, getQueryParams } from '../../../../helpers/utils';
import { RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import GenericPlansReducer, {
  genericFetchPlans,
  GenericPlan,
  makeGenericPlansArraySelector,
  reducerName as genericReducerName,
} from '../../../../store/ducks/generic/plans';
import { InterventionType } from '../../../../store/ducks/plans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the MDA plan definitions reducer */
reducerRegistry.register(genericReducerName, GenericPlansReducer);

/** a list of plan statuses to be displayed */
const allowedPlanStatusList = getPlanStatusToDisplay(HIDDEN_PLAN_STATUSES);

/** selector for MDA point plans */
const makeMDAPointPlansArraySelector = makeGenericPlansArraySelector();

/** Simple component that loads a preview list of MDA plans */
const MDAPointPlansList = (props: GenericPlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for MDAPointPlansList */
const defaultProps: GenericPlanListProps = {
  fetchPlans: genericFetchPlans,
  pageTitle: MDA_POINT_PLANS,
  pageUrl: REPORT_MDA_POINT_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_MDA_POINT_REPORTING_PLANS_SLICE,
};

MDAPointPlansList.defaultProps = defaultProps;

export { MDAPointPlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: GenericPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const MDAPointPlansArray = makeMDAPointPlansArraySelector(state, {
    interventionTypes: [InterventionType.MDAPoint],
    plan_title: searchedTitle,
    statusList: allowedPlanStatusList,
  });

  return {
    ...ownProps,
    plans: MDAPointPlansArray,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: genericFetchPlans };

/** Connected ActiveFI component */
const ConnectedMDAPointPlansList = connect(mapStateToProps, mapDispatchToProps)(MDAPointPlansList);

export default ConnectedMDAPointPlansList;
