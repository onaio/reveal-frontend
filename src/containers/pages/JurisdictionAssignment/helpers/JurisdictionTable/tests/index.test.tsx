import { Dictionary } from '@onaio/utils/dist/types/types';
import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { Route, Router, Switch } from 'react-router';
import { MemoryRouter, RouteComponentProps } from 'react-router-dom';
import { ConnectedJurisdictionTable } from '..';
import { ASSIGN_PLAN_URL, MANUAL_ASSIGN_JURISDICTIONS_URL } from '../../../../../../constants';
import store from '../../../../../../store';
import {
  deforest,
  deselectAllNodes,
  fetchTree,
} from '../../../../../../store/ducks/opensrp/hierarchies';
import { sampleHierarchy } from '../../../../../../store/ducks/opensrp/hierarchies/tests/fixtures';
import {
  irsPlans,
  plans,
} from '../../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { PlanStatus } from '../../../../../../store/ducks/plans';
import { afterDraftSave, afterSaveAndActivate } from './fixtures';

// tslint:disable-next-line: no-var-requires
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

jest.mock('reactstrap', () => {
  const original = require.requireActual('reactstrap');
  return { ...original, Tooltip: () => null };
});

/** will use this to render table rows */
const renderTable = (wrapper: ReactWrapper, message: string) => {
  const trs = wrapper.find('table tr');
  trs.forEach(tr => {
    expect(tr.text()).toMatchSnapshot(message);
  });
};

describe('src/containers/pages/jurisdictionView/jurisdictionTable', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    fetch.resetMocks();
  });

  beforeEach(() => {
    store.dispatch(fetchTree(sampleHierarchy));
  });

  afterEach(() => {
    store.dispatch(deselectAllNodes('2942'));
    store.dispatch(deforest());
  });

  it('works correctly through a full render cycle', async () => {
    const plan = irsPlans[0];

    const activateDiv = document.createElement('div');
    activateDiv.setAttribute('id', 'save-and-activate-wrapper');
    document.body.appendChild(activateDiv);

    const draftDiv = document.createElement('div');
    draftDiv.setAttribute('id', 'save-draft-wrapper');
    document.body.appendChild(draftDiv);

    /** current architecture does not use the Jurisdiction table as a view
     * it is seen as a controlled component that is feed some data from controlling component
     * That's why there was a need to have this custom view and a mock routing system.
     */
    const CustomView = (props: RouteComponentProps<Dictionary>) => {
      const mockProps = {
        autoSelectionFlow: false,
        currentParentId: props.match.params.parentId,
        plan,
        rootJurisdictionId: '2942',
      };
      return <ConnectedJurisdictionTable {...mockProps} />;
    };

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={CustomView}
          />
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId`}
            component={CustomView}
          />
        </Switch>
      );
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${MANUAL_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    renderTable(wrapper, 'should have single node(parent node)');
    const tbodyRow = wrapper.find('tbody tr');
    expect(tbodyRow.length).toEqual(1);
    expect(
      tbodyRow
        .at(0)
        .text()
        .includes('Lusaka')
    ).toBeTruthy();

    // simulate click on checkbox to check
    tbodyRow.find('input').simulate('change', { target: { name: '', checked: true } });
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });
    renderTable(wrapper, 'input checked');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
        .includes('1')
    ).toBeTruthy();

    // test drilldown
    tbodyRow
      .at(0)
      .find('NodeCell Link a')
      .simulate('click', { button: 0 });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });
    renderTable(wrapper, 'after first click');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
        .includes('Mtendere')
    ).toBeTruthy();

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });
    renderTable(wrapper, 'after second click');
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
        .includes('Akros_1')
    ).toBeTruthy();

    // akros is the last child it should not be clickable
    expect(toJson(wrapper.find('tbody tr NodeCell span'))).toMatchSnapshot(
      'should not have btn-link or onClick handler'
    );
  });

  it('selects and deselect nodes', async () => {
    const plan = irsPlans[0];
    const props = {
      plan,
      rootJurisdictionId: '2942',
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedJurisdictionTable {...props} />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });
    // simulate click on parent checkbox to check
    wrapper
      .find('.thead-plan-orgs input')
      .simulate('change', { target: { name: '', checked: true } });

    wrapper.update();
    expect(toJson(wrapper.find('.thead-plan-orgs input'))).toMatchSnapshot('should now be checked');

    // simulate click on parent checkbox to uncheck
    wrapper
      .find('.thead-plan-orgs input')
      .simulate('change', { target: { name: '', checked: false } });

    wrapper.update();
    expect(toJson(wrapper.find('.thead-plan-orgs input'))).toMatchSnapshot('should be unchecked');

    const parentNodeRow = wrapper.find('tbody tr').at(0);
    // create snapshot of checkbox before getting checked
    expect(toJson(parentNodeRow.find('input'))).toMatchSnapshot('should be unchecked');

    // simulate click on checkbox to check
    parentNodeRow.find('input').simulate('change', { target: { name: '', checked: true } });
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(
      toJson(
        wrapper
          .find('tbody tr')
          .at(0)
          .find('input')
      )
    ).toMatchSnapshot('should be now checked');

    // simulate click on checkbox to unchecked
    wrapper
      .find('tbody tr')
      .at(0)
      .find('input')
      .simulate('change', { target: { name: '', checked: false } });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // checkbox is now deselected again
    expect(
      toJson(
        wrapper
          .find('tbody tr')
          .at(0)
          .find('input')
      )
    ).toMatchSnapshot('should be now deselected');
  });

  it('disables and enables checkbox for FI plans', async () => {
    const plan = plans[0];
    const CustomView = (props: RouteComponentProps<Dictionary>) => {
      const mockProps = {
        currentParentId: props.match.params.parentId,
        plan,
        rootJurisdictionId: '2942',
      };
      return <ConnectedJurisdictionTable {...mockProps} />;
    };

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={CustomView}
          />
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId`}
            component={CustomView}
          />
        </Switch>
      );
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${MANUAL_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    const tbodyRow = wrapper.find('tbody tr');
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .find('input')
        .prop('disabled')
    ).toBe(true);

    // drilldown
    tbodyRow
      .at(0)
      .find('NodeCell Link a')
      .simulate('click', { button: 0 });
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .find('input')
        .prop('disabled')
    ).toBe(true);

    // drill down
    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .find('input')
        .prop('disabled')
    ).toBe(false);
  });

  it('makes correct calls after clicking save draft action', async () => {
    const plan = { ...plans[0], status: PlanStatus.DRAFT };

    const CustomView = (props: RouteComponentProps<Dictionary>) => {
      const mockProps = {
        currentParentId: props.match.params.parentId,
        plan,
        rootJurisdictionId: '2942',
      };
      return <ConnectedJurisdictionTable {...mockProps} />;
    };

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={CustomView}
          />
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId`}
            component={CustomView}
          />
        </Switch>
      );
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${MANUAL_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // save draft button is disabled
    expect(wrapper.find('#save-draft button').prop('disabled')).toEqual(true);
    // select a jurisdiction
    const tbodyRow = wrapper.find('tbody tr');
    tbodyRow.find('input').simulate('change', { target: { name: '', checked: true } });
    // save draft button is enabled
    expect(wrapper.find('#save-draft button').prop('disabled')).toEqual(false);

    // simulate a click on draft
    wrapper.find('#save-draft button').simulate('click');

    // flush promises
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // check fetch request
    expect(fetch.mock.calls[0]).toEqual(afterDraftSave);
  });

  it('makes correct calls after clicking save and activate', async () => {
    const plan = { ...plans[0], status: PlanStatus.DRAFT };

    const CustomView = (props: RouteComponentProps<Dictionary>) => {
      const mockProps = {
        currentParentId: props.match.params.parentId,
        plan,
        rootJurisdictionId: '2942',
      };
      return <ConnectedJurisdictionTable {...mockProps} />;
    };

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={CustomView}
          />
          <Route
            exact={true}
            path={`${MANUAL_ASSIGN_JURISDICTIONS_URL}/:planId`}
            component={CustomView}
          />
        </Switch>
      );
    };

    fetch.once(JSON.stringify(sampleHierarchy), { status: 200 });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${MANUAL_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    // first flush promises
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // save & activate button is disabled
    expect(wrapper.find('#save-and-activate button').prop('disabled')).toEqual(true);
    // select a jurisdiction
    const tbodyRow = wrapper.find('tbody tr');
    tbodyRow.find('input').simulate('change', { target: { name: '', checked: true } });
    // save & activate button is enabled
    expect(wrapper.find('#save-and-activate button').prop('disabled')).toEqual(false);

    // simulate a click on draft
    wrapper.find('#save-and-activate button').simulate('click');

    // flush promises
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // check fetch request
    expect(fetch.mock.calls[0]).toEqual(afterSaveAndActivate);

    // check redirection
    expect((wrapper.find('Router').props() as any).history.location.pathname).toEqual(
      `${ASSIGN_PLAN_URL}/${plan.identifier}`
    );
  });
});
