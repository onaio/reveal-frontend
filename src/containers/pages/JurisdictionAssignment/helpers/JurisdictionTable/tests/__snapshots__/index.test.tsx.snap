// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/containers/pages/jurisdictionView/jurisdictionTable selects and deselect nodes: should be now checked 1`] = `
<input
  checked={true}
  disabled={false}
  key="2942-check-jurisdiction"
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable selects and deselect nodes: should be now deselected 1`] = `
<input
  checked={false}
  disabled={false}
  key="2942-check-jurisdiction"
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable selects and deselect nodes: should be unchecked 1`] = `
<input
  checked={false}
  disabled={false}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable selects and deselect nodes: should be unchecked 2`] = `
<input
  checked={false}
  disabled={false}
  key="2942-check-jurisdiction"
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable selects and deselect nodes: should now be checked 1`] = `
<input
  checked={true}
  disabled={false}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: after first click 1`] = `"NameStructures CountTarget StatusStatus setting by...Selected Jurisdictions"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: after first click 2`] = `"Mtendere159User Change1"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: after second click 1`] = `"NameStructures CountTarget StatusStatus setting by...Selected Jurisdictions"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: after second click 2`] = `"Akros_1159TargetedUser Change1"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: input checked 1`] = `"NameStructures CountTarget StatusStatus setting by...Selected Jurisdictions"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: input checked 2`] = `"Lusaka159User Change1"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: should have single node(parent node) 1`] = `"NameStructures CountTarget StatusStatus setting by...Selected Jurisdictions"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: should have single node(parent node) 2`] = `"Lusaka1590"`;

exports[`src/containers/pages/jurisdictionView/jurisdictionTable works correctly through a full render cycle: should not have btn-link or onClick handler 1`] = `
<span
  key="3951-span"
>
  Akros_1
</span>
`;
